sdk/identity/azure-identity/azure/identity/_credentials/win_vscode_credential.py

```diff
@@ -0,0 +1,155 @@
+)
+from .._internal.aad_client import AadClient
+try:
+    import ctypes.wintypes as wt
+except (IOError, ValueError):
+    pass
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Iterable, Optional
+    from azure.core.credentials import AccessToken
+
+SUPPORTED_CREDKEYS = set((
+    'Type', 'TargetName', 'Persist',
+    'UserName', 'Comment', 'CredentialBlob'))
+
+
+class _CREDENTIAL(ct.Structure):
+    _fields_ = [
+        ("Flags", wt.DWORD),
+        ("Type", wt.DWORD),
+        ("TargetName", ct.c_wchar_p),
+        ("Comment", ct.c_wchar_p),
+        ("LastWritten", wt.FILETIME),
+        ("CredentialBlobSize", wt.DWORD),
+        ("CredentialBlob", wt.LPBYTE),
+        ("Persist", wt.DWORD),
+        ("AttributeCount", wt.DWORD),
+        ("Attributes", ct.c_void_p),
+        ("TargetAlias", ct.c_wchar_p),
+        ("UserName", ct.c_wchar_p)]
+
+    @classmethod
+    def from_dict(cls, credential):
+        # pylint:disable=attribute-defined-outside-init
+        creds = cls()
+        pcreds = _PCREDENTIAL(creds)
+
+        ct.memset(pcreds, 0, ct.sizeof(creds))
+
+        for key in SUPPORTED_CREDKEYS:
+            if key in credential:
+                if key != 'CredentialBlob':
+                    setattr(creds, key, credential[key])
+                else:
+                    blob = credential['CredentialBlob']
+                    blob_data = ct.create_unicode_buffer(blob)
+                    creds.CredentialBlobSize = \
+                        ct.sizeof(blob_data) - \
+                        ct.sizeof(ct.c_wchar)
+                    creds.CredentialBlob = ct.cast(blob_data, wt.LPBYTE)
+        return creds
+
+
+_PCREDENTIAL = ct.POINTER(_CREDENTIAL)
+
+
+_advapi = ct.WinDLL('advapi32')
+_advapi.CredWriteW.argtypes = [_PCREDENTIAL, wt.DWORD]
+_advapi.CredWriteW.restype = wt.BOOL
+_advapi.CredReadW.argtypes = [wt.LPCWSTR, wt.DWORD, wt.DWORD, ct.POINTER(_PCREDENTIAL)]
+_advapi.CredReadW.restype = wt.BOOL
+_advapi.CredFree.argtypes = [_PCREDENTIAL]
+_advapi.CredDeleteW.restype = wt.BOOL
+_advapi.CredDeleteW.argtypes = [wt.LPCWSTR, wt.DWORD, wt.DWORD]
+
+
+def _cred_write(credential):
+    creds = _CREDENTIAL.from_dict(credential)
+    cred_ptr = _PCREDENTIAL(creds)
+    _advapi.CredWriteW(cred_ptr, 0)
+
+
+def _cred_delete(service_name, account_name):
+    target = u"{}/{}".format(service_name, account_name)
+    _advapi.CredDeleteW(target, 1, 0)
+
+
+def _read_credential(service_name, account_name):
+    target = u"{}/{}".format(service_name, account_name)
+    cred_ptr = _PCREDENTIAL()
+    if _advapi.CredReadW(target, 1, 0, ct.byref(cred_ptr)):
+        cred_blob = cred_ptr.contents.CredentialBlob
+        cred_blob_size = cred_ptr.contents.CredentialBlobSize
+        password_as_list = [int.from_bytes(cred_blob[pos:pos + 2], 'little')
+                            for pos in range(0, cred_blob_size, 2)]
+        cred = ''.join(map(chr, password_as_list))
```

Comment:
I think these three lines could just be `cred = cred_blob.decode("utf-16-le")`, right?

Reply:
I tried both cred = str(cred_blob, "utf-16") & cred = str(cred_blob, "utf-16-le"), neither of them worked correctly.

Reply:
Looking at the value on my machine for this, they're storing it as ASCII, not UTF-16. (Specifically, it looks like URL-safe base64 encoding without padding, but that probably doesn't matter as I think you just use this directly, right?)

Reply:
Yes. In our case, we use it directly.